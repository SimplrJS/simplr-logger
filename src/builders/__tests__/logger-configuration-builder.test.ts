import { LoggerConfigurationBuilder, LogLevel, ConsoleMessageHandler } from "simplr-logger";

test("Default Configuration", () => {
    const config = new LoggerConfigurationBuilder().Build();
    expect(config).toMatchSnapshot();
});

test("Initial Configuration", () => {
    const prefix = new LoggerConfigurationBuilder({
        Prefix: "[SimplrLogger prefix]"
    }).Build();
    expect(prefix).toMatchSnapshot();

    const logLevel = new LoggerConfigurationBuilder({
        DefaultLogLevel: {
            LogLevel: LogLevel.Critical,
            LogLevelIsBitMask: false
        }
    }).Build();
    expect(logLevel).toMatchSnapshot();

    const logLevelBitMask = new LoggerConfigurationBuilder({
        DefaultLogLevel: {
            LogLevel: LogLevel.Critical | LogLevel.Warning | LogLevel.Error,
            LogLevelIsBitMask: true
        }
    }).Build();
    expect(logLevelBitMask).toMatchSnapshot();

    const handlers = new LoggerConfigurationBuilder({
        WriteMessageHandlers: [
            { Handler: new ConsoleMessageHandler() },
            { Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Trace },
            { Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Trace, LogLevelIsBitMask: false },
            { Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Critical | LogLevel.Error, LogLevelIsBitMask: true }
        ]
    }).Build();
    expect(handlers).toMatchSnapshot();
});

describe("Configuration Methods", () => {
    test("SetPrefix", () => {
        const prefixConfig = new LoggerConfigurationBuilder()
            .SetPrefix("[SimplrLogger prefix]")
            .Build();

        expect(prefixConfig).toMatchSnapshot();
    });

    test("SetDefaultLogLevel", () => {
        const logLevelConfig = new LoggerConfigurationBuilder()
            .SetDefaultLogLevel(LogLevel.Trace)
            .Build();

        expect(logLevelConfig).toMatchSnapshot();

        const logLevelBitMaskConfig = new LoggerConfigurationBuilder()
            .SetDefaultLogLevel([LogLevel.Trace, LogLevel.Information, LogLevel.Warning])
            .Build();

        expect(logLevelBitMaskConfig).toMatchSnapshot();
    });

    test("AddWriteMessageHandler", () => {
        const config = new LoggerConfigurationBuilder()
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler() })
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler() }, LogLevel.Information)
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Critical })
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Critical }, LogLevel.None)
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Error, LogLevelIsBitMask: false })
            .AddWriteMessageHandler({
                Handler: new ConsoleMessageHandler(),
                LogLevel: LogLevel.Error, LogLevelIsBitMask: false
            }, LogLevel.None)
            .AddWriteMessageHandler({
                Handler: new ConsoleMessageHandler(),
                LogLevel: LogLevel.Debug | LogLevel.Critical, LogLevelIsBitMask: true
            })
            .AddWriteMessageHandler({
                Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Debug | LogLevel.Critical,
                LogLevelIsBitMask: true
            }, LogLevel.None)
            .Build();

        expect(config).toMatchSnapshot();
    });

    test("AddWriteMessageHandlers", () => {
        const config = new LoggerConfigurationBuilder()
            .AddWriteMessageHandlers([{ Handler: new ConsoleMessageHandler() }])
            .AddWriteMessageHandlers([
                { Handler: new ConsoleMessageHandler() },
                { Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Debug }
            ])
            .Build();

        expect(config).toMatchSnapshot();
    });

    test("Override", () => {
        const config = new LoggerConfigurationBuilder()
            .SetPrefix("Start prefix")
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Debug })
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Trace })
            .SetDefaultLogLevel(LogLevel.Trace)
            .Override({ Prefix: "New Prefix" })
            .Build();

        expect(config).toMatchSnapshot();

        const config2 = new LoggerConfigurationBuilder()
            .SetPrefix("Start prefix")
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Debug })
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Trace })
            .SetDefaultLogLevel(LogLevel.Trace)
            .Override({ WriteMessageHandlers: [] })
            .Build();

        expect(config2).toMatchSnapshot();

        const config3 = new LoggerConfigurationBuilder()
            .SetPrefix("Start prefix")
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Debug })
            .AddWriteMessageHandler({ Handler: new ConsoleMessageHandler(), LogLevel: LogLevel.Trace })
            .SetDefaultLogLevel(LogLevel.Trace)
            .Override({ DefaultLogLevel: { LogLevel: LogLevel.None | LogLevel.Error | LogLevel.Warning, LogLevelIsBitMask: true } })
            .Build();

        expect(config3).toMatchSnapshot();
    });
});
