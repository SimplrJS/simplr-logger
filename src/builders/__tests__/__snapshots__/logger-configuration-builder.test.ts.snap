// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration Methods AddWriteMessageHandler 1`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 4,
    "LogLevelIsBitMask": false,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 4,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 8,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 1,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 1,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 2,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 2,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 17,
      "LogLevelIsBitMask": true,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 17,
      "LogLevelIsBitMask": true,
    },
  ],
}
`;

exports[`Configuration Methods AddWriteMessageHandlers 1`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 4,
    "LogLevelIsBitMask": false,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 4,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 4,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 16,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Configuration Methods Override 1`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 32,
    "LogLevelIsBitMask": false,
  },
  "Prefix": "New Prefix",
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 16,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 32,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Configuration Methods Override 2`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 32,
    "LogLevelIsBitMask": false,
  },
  "Prefix": "Start prefix",
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 32,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Configuration Methods Override 3`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 6,
    "LogLevelIsBitMask": true,
  },
  "Prefix": "Start prefix",
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 16,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 32,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Configuration Methods SetDefaultLogLevel 1`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 32,
    "LogLevelIsBitMask": false,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 32,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Configuration Methods SetDefaultLogLevel 2`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 44,
    "LogLevelIsBitMask": true,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 44,
      "LogLevelIsBitMask": true,
    },
  ],
}
`;

exports[`Configuration Methods SetPrefix 1`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 4,
    "LogLevelIsBitMask": false,
  },
  "Prefix": "[SimplrLogger prefix]",
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 4,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Default Configuration 1`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 4,
    "LogLevelIsBitMask": false,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 4,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Initial Configuration 1`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 4,
    "LogLevelIsBitMask": false,
  },
  "Prefix": "[SimplrLogger prefix]",
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 4,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Initial Configuration 2`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 1,
    "LogLevelIsBitMask": false,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 1,
      "LogLevelIsBitMask": false,
    },
  ],
}
`;

exports[`Initial Configuration 3`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 7,
    "LogLevelIsBitMask": true,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 7,
      "LogLevelIsBitMask": true,
    },
  ],
}
`;

exports[`Initial Configuration 4`] = `
Object {
  "DefaultLogLevel": Object {
    "LogLevel": 4,
    "LogLevelIsBitMask": false,
  },
  "Prefix": undefined,
  "WriteMessageHandlers": Array [
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 4,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 32,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 32,
      "LogLevelIsBitMask": false,
    },
    Object {
      "Handler": ConsoleMessageHandler {
        "configuration": Object {
          "LogLevelPrefix": "Short",
          "TimePrefix": "Short",
          "UseColors": false,
        },
      },
      "LogLevel": 3,
      "LogLevelIsBitMask": true,
    },
  ],
}
`;
